kroxylicious/src/main/java/io/kroxylicious/proxy/internal/filter/UpstreamBrokerAddressCachingNetFilter.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.kroxylicious.proxy.internal.filter;
+
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import io.kroxylicious.proxy.bootstrap.FilterChainFactory;
+import io.kroxylicious.proxy.filter.KrpcFilter;
+import io.kroxylicious.proxy.filter.MetadataResponseFilter;
+import io.kroxylicious.proxy.filter.NetFilter;
+import io.kroxylicious.proxy.service.ClusterEndpointProvider;
+import io.kroxylicious.proxy.service.HostPort;
+
+/**
+ * Temporary implementation of NetFilter that uses the port number from the local end of the
+ * downstream connection to identify the virtual clusters broker.  It then uses this to identify
+ * the upstream broker address from a cache built from a previous connection's metadata response.
+ */
+public class UpstreamBrokerAddressCachingNetFilter implements NetFilter {
+
+    private final HostPort targetClusterBootstrap;
+    private final FilterChainFactory filterChainFactory;
+    private final ClusterEndpointProvider endpointProvider;
+
+    private final Map<Integer, HostPort> upstreamBrokers = new ConcurrentHashMap<>();
+
+    public UpstreamBrokerAddressCachingNetFilter(HostPort targetClusterBootstrap, FilterChainFactory filterChainFactory,
+                                                 ClusterEndpointProvider endpointProvider) {
+        this.targetClusterBootstrap = targetClusterBootstrap;
+        this.filterChainFactory = filterChainFactory;
+        this.endpointProvider = endpointProvider;
+    }
+
+    @Override
+    public void selectServer(NetFilterContext context) {
+        var filters = new ArrayList<>(Arrays.stream(filterChainFactory.createFilters()).toList());
+
+        // Add a filter to the *end of the chain* that gathers the true nodeId/upstream broker mapping.
+        filters.add((MetadataResponseFilter) (header, response, filterContext) -> {
+            response.brokers().forEach(b -> upstreamBrokers.put(b.nodeId(), new HostPort(b.host(), b.port())));
+            filterContext.forwardResponse(response);
+        });
+
+        var endpointMatchResult = endpointProvider.hasMatchingEndpoint(context.sniHostname(),
+                ((InetSocketAddress) context.localAddress()).getPort());
+
+        HostPort target;
+        if (endpointMatchResult.matched() && endpointMatchResult.nodeId() != null) {
+            target = upstreamBrokers.getOrDefault(endpointMatchResult.nodeId(), targetClusterBootstrap);
+        }
+        else {
+            target = targetClusterBootstrap;
```

Comment:
Naively I would expect `endpointMatchResult` to be `true` for a bootstrap address, but I guess this is really the else clause for `nodeId == null`. 

Oh the [findsBootstrapMatchByPort](https://github.com/kroxylicious/kroxylicious/pull/254/files#diff-3f8c7760c1b6d2140396772ff3302a65927bcf0b5ecf445a77a54147ee31840aR66) test backs up that assumption. Which makes me wonder if this is the correct behaviour when the `endpointMatchResult` is `false`?  Should we reject the connection attempt? What does it mean from a security perspective?

Reply:
> Naively I would expect `endpointMatchResult` to be true for a bootstrap address, 

That was my intent.  Isn't that what I did?  I don't see it.

> Oh the [findsBootstrapMatchByPort](https://github.com/kroxylicious/kroxylicious/pull/254/files#diff-3f8c7760c1b6d2140396772ff3302a65927bcf0b5ecf445a77a54147ee31840aR66) test backs up that assumption. Which makes me wonder if this is the correct behaviour when the endpointMatchResult is false? Should we reject the connection attempt? What does it mean from a security perspective?

The behaviour in the PR definitely wrong.  It potentially lets a client get a connection to the wrong broker.

I've been thinking about this case over the weekend.  I think we'll see this when a kroxy is restarted. Clients  may try to reconnect their existing connections to the brokers.  In this case, kroxy won't yet know the cluster topology so it won't know how to connect to broker-n on the upstream side. It only knowns bootstrap.   I think in this case, the correct thing to do it is 'hijack' the connection: kroxy makes the connection to bootstrap, lets the client authenticate then emits its own MetadataRequest.  That furnishes kroxy with the topology. It *must* then close the connection - otherwise we run the risk that the client continues connected to the wrong broker.  The client will retry its connection to broker-n, and this time it is work.

I'm not proposing making this change now. 

I think the code as it stands is good enough for an interim step, but I'll add a comment to make sure this isn't forgotten.